gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java

```diff
@@ -296,7 +297,14 @@ private HttpClientRequest prepareRequest(final HttpServerRequest req, final Stri
 
             req.response().setStatusCode(statusCode);
 
-            req.response().headers().addAll(cRes.headers());
+            // Add received headers to original request except the 'Connection'
+            // headers. See RFC "https://tools.ietf.org/html/rfc2616#section-14.10".
+            // HINT: This manipulates the originally received headers.
+            final MultiMap headersFromServer = cRes.headers();
+            headersFromServer.getAll("connection").forEach(headersFromServer::remove);
```

Comment:
I think it's implemented a little bit complicated. I made a short unit test:

```java
@Test
public void testMultimap() {
	CaseInsensitiveHeaders params = new CaseInsensitiveHeaders();

	params.add("connection", "abc");
	params.add("test", "123");
	params.add("connection", "def");

	System.out.println("names: " + params.names());
	System.out.println("getall connection: " + params.getAll("connection"));

	params.remove("connection");

	System.out.println("getall connection after remove: " + params.getAll("connection"));
	System.out.println("names after remove: " + params.names());
}
```
Output:
```
names: [connection, test]
getall connection: [abc, def]
getall connection after remove: []
names after remove: [test]
```
So it looks like it could be simply implemented with
```java
req.response().headers().addAll(cRes.headers().remove("connection"));
```
Has to be tested :-)

Reply:
This would not fulfill the specification (see: https://tools.ietf.org/html/rfc2616#section-14.10)

> HTTP/1.1 proxies MUST parse the Connection header field before a
   message is forwarded and, for each connection-token in this field,
   remove any header field(s) from the message with the same name as the
   connection-token. Connection options are signaled by the presence of
   a connection-token in the Connection header field

For example we receiving following headers:

> Connection: close, foo
> Connection: bar, dada
> foo: blubb
> bar: asdf

In this situation we've to remove *all* of these 4 headers.

Reply:
You're right. For the request headers it already was implmented like this. Maybe make a private method to remove these headers, to reduce code duplication